src/coreclr/jit/lowerxarch.cpp

```diff
@@ -656,82 +655,437 @@ void Lowering::LowerPutArgStk(GenTreePutArgStk* putArgStk)
+        {
+            lowerBound = comp->gtNewSimdZeroNode(TYP_SIMD16, simdBaseJitType, 16, /* isSimdAsHWIntrinsic */ false);
+            BlockRange().InsertAfter(castOp, lowerBound);
+            LowerNode(lowerBound);
+        }
+        else
+        {
+            GenTree* lowerBoundCns;
+
+            if (castToType == TYP_BYTE)
+            {
+                lowerBoundCns = comp->gtNewDconNode(-128.0, castFromType);
+            }
+            else
+            {
+                assert(castToType == TYP_SHORT);
+                lowerBoundCns = comp->gtNewDconNode(-32768.0, castFromType);
+            }
+
+            BlockRange().InsertAfter(castOp, lowerBoundCns);
+
+            lowerBound = comp->gtNewSimdHWIntrinsicNode(TYP_SIMD16, lowerBoundCns, NI_Vector128_CreateScalarUnsafe,
+                                                        simdBaseJitType, 16);
+            BlockRange().InsertAfter(lowerBoundCns, lowerBound);
+            LowerNode(lowerBound);
+        }
+
+        GenTree* simdMax =
+            comp->gtNewSimdHWIntrinsicNode(TYP_SIMD16, lowerBound, castOp, simdMaxIntrinId, simdBaseJitType, 16);
+        BlockRange().InsertAfter(lowerBound, simdMax);
+        LowerNode(simdMax);
+
+        // We min ourselves against the upper boun
+        GenTree* upperBound;
+        GenTree* upperBoundCns;
+
+        switch (castToType)
+        {
+            case TYP_BYTE:
+            {
+                upperBoundCns = comp->gtNewDconNode(+127.0, castFromType);
+                break;
+            }
+
+            case TYP_UBYTE:
+            {
+                upperBoundCns = comp->gtNewDconNode(+255.0, castFromType);
+                break;
+            }
+
+            case TYP_SHORT:
+            {
+                upperBoundCns = comp->gtNewDconNode(+32767.0, castFromType);
+                break;
+            }
+
+            case TYP_USHORT:
+            {
+                upperBoundCns = comp->gtNewDconNode(+65535.0, castFromType);
+                break;
+            }
+
+            default:
+            {
+                unreached();
+            }
+        }
+
+        BlockRange().InsertAfter(simdMax, upperBoundCns);
+
+        upperBound = comp->gtNewSimdHWIntrinsicNode(TYP_SIMD16, upperBoundCns, NI_Vector128_CreateScalarUnsafe,
+                                                    simdBaseJitType, 16);
+        BlockRange().InsertAfter(upperBoundCns, upperBound);
+        LowerNode(upperBound);
+
+        GenTree* simdMin =
+            comp->gtNewSimdHWIntrinsicNode(TYP_SIMD16, upperBound, simdMax, simdMinIntrinId, simdBaseJitType, 16);
+        BlockRange().InsertAfter(upperBound, simdMin);
+        LowerNode(simdMin);
+
+        // Replace the castOp with the saturated value
+        castOp = comp->gtNewSimdHWIntrinsicNode(castFromType, simdMin, NI_Vector128_ToScalar, simdBaseJitType, 16);
+        BlockRange().InsertAfter(simdMin, castOp);
+        LowerNode(castOp);
+
         tmpType = TYP_INT;
```

Comment:
Why do we need an intermediate cast here?

Reply:
Same as above, the `tmpType` needs to be `int` because the underlying instruction produces a `TYP_INT`.

Reply:
Seeing as this logic is identical to the one in `lowerarmarch.cpp`, I presume https://github.com/dotnet/runtime/pull/61761#discussion_r753727709 applies this it as well.

Reply:
Resolved as per https://github.com/dotnet/runtime/pull/61761#discussion_r753727709.